### PR TITLE
Topic recognition (Siamese ADNI) - s4823870

### PR DESCRIPTION
**Pull request by rachitchaurasia**
# Classifier Based on Siamese Network for Alzheimer's Disease Classification

## Question
I chose task 7 to create a classifier based on a Siamese network to classify Alzheimer's disease (AD) and normal brain data from the ADNI dataset. The goal is to achieve an accuracy of approximately 0.8 on the test set. I've already implemented code for a Siamese network, dataset loading, training, and classification.

## Context
I have provided the following code files and additional details:

**1. modules.py:** Contains the implementation of the Siamese network and the contrastive loss used for training the network.

**2. dataset.py:** Defines custom datasets for Siamese network training and image classification, including data loading functions.

**3. train.py:** The script for training a Siamese network for image similarity learning. It uses the Siamese network from `modules.py`, the custom dataset from `dataset.py`, and contrastive loss.

**4. predict.py:** The script for image classification using a pre-trained Siamese network. It also includes a classifier for classification based on Siamese network embeddings.

**5. Folders:**
   - `result-plot`: Contains generated plots.
   - `assets`: Contains assets for the `readme.md` file.
   - `requirements.txt`: The requirements file for the project.

I've already provided header blocks for each of the code files with my name and student ID. All files are functional.

Please review the provided code and suggest improvements or feedback wherever needed. Any guidance, optimizations, or modifications are greatly appreciated.

## Additional Information
- The input data is from the ADNI dataset, and the task is to classify between Alzheimer's disease (AD) and normal (NC) cases.
- The Siamese network architecture and loss functions have been implemented.
- The `train.py` script trains the Siamese network and saves the trained model.
- The `predict.py` script uses a pre-trained Siamese network to classify images.
- I'm using PyTorch for this project.

## Files
You can find the code files in the respective sections above.

Thank you for your assistance!

_Rachit Chaurasia (@rachitchaurasia)
UQ ID: s4823870_

### CODE REVIEW

**Comment by nathasha-naranpanawa** [comment]
**This is an initial inspection, no action is required at this point**

Difficulty: Hard

Readme: Good
- Clear flow of information
- Algortihm and raining choices explained well
- minimal plots provided for only the siamese network, they seem to be only for the training set (not very useful by itself)
- no loss/accuracy plots or metrics for the classifier

Commit messages: OK

Code: 
- design is somewhat poor
- classifier is trained only during prediction. 
- although required 4 python files are present, code components within them are not planned well and included somewhat incorrectly (eg: training the classifier should be performed within train.py, not in predict.py, classifier network should be defined in modules.py etc.)
- sufficiently commented

Functionality/Performance: 
- **no average classification accuracy numbers provided for the test data**
- loss plots included doesn't help judge the performance as they seem to be only for the training set

General comments:
- **algorithm solves the respective problem somewhat half-way**
- you need to have trained both the siamese and classifier and saved the weights (instead of training the classifier separately for prediction only), and written the code for class prediction of an image using saved weights in predict.py

**Comment by shakes76** [comment]
# Marking
## Good Practice (Design/Commenting, TF/Torch Usage)
Adequate design and implementation
Good spacing and comments
Header blocks
## Recognition Problem 
Partially Solves problem, no classifier in training -3
Driver Script present
File structure present
Shows Usage & Demo & Visualisation & Data usage, loss plots not provided -1
Module present
Commenting
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
ReadMe very good
Model/technical explanation
Good Description and Comments 
Markdown used and PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch) 
No Feedback required
Request Description good

**Comment by shakes76** [comment]
Remove gitignore file so I can merge. does not affect grade only the merge

**Comment by rachitchaurasia** [comment]
> Remove gitignore file so I can merge. does not affect grade only the merge

Removed .gitignore file for merging.

**Comment by wangzhaomxy** [comment]
Feedback attempt and no feedback marks lost.
